Add default permissions to admins and operators

shashwat1002 opened the pull request:
fixes issue #48
The consequence of this code is that operators have permissions in accordance with the specifications mentioned in the issue and the admins have access to all the objects mentioned in the issue.

atb00ker commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,66 @@
+    
+
+class Migration(migrations.Migration): 
+    dependencies=[
```
dependencies are not correct. Please fix this. :smile:

atb00ker commented on .travis.yml:
```diff
@@ -51,4 +51,4 @@ script:
 
 after_success:
-  pipenv run coveralls
+  pipenv run coveralls
```
Please add a line after this line.

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,66 @@
+
+    operations=[
+        migrations.RunPython(make_default_permissions_in_code),
+        migrations.RunPython(assignPerm),
```
add reverse_code=migrations.RunPython.noop as it's done here: https://github.com/openwisp/openwisp-users/blob/master/openwisp_users/migrations/0004_default_groups.py

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,66 @@
+
+
+
+
```
leave only 1 final blank line

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,66 @@
+from django.contrib.auth.models import Permission
+from django.contrib.auth.management import create_permissions 
+from django.db import transaction
+
```
add 1 blank line as per PEP8

shashwat1002 commented:
Squashed all my changes into the final commit. 
Made the required stylistic changes in accordance with PEP8. 
Changed the names of certain functions to make them more expressive.

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,64 @@
+        for operation in manage_operations:
+            permission = Permission.objects.get(
+                codename='{}_{}'.format(operation, modelClass)
+                )
```
the indentation style is weird, use the following please:

```python
permission = Permission.objects.get(
    codename='{}_{}'.format(operation, modelClass)
)
```

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,64 @@
+        ('pki', '__first__'),
+        ('geo', '__first__'),
+        ('config', '0014_device_hardware_id'),
+
```
remove blank line

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,64 @@
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ('openwisp_users', '__first__'),
```
this shouldn't be necessary, try to remove it, if tests pass after you remove it don't add it back please

atb00ker replied:
@shashwat1002 this is remaining!

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,64 @@
+    dependencies = [
+        ('openwisp_users', '__first__'),
+        ('openwisp_users', '0004_default_groups'),
+        ('pki', '__first__'),
```
this shouldn't be necessary, try to remove it, if tests pass after you remove it don't add it back please

shashwat1002 replied:
I tried migrating without this: dosen't work because those migrations need to run before I create the permissions, so that my function can access all those models. Since, I'm creating the permissions in this particular file (I'm doing this because this is the first app of openwisp_controller to be migrated), I need to put that into the dependencies. 
If I don't, I get the 
`django.contrib.auth.models.DoesNotExist: Permission matching query does not exist.`
my understanding of this error is that django dosen't create the permissions for these  models as those migrations haven't run yet and consequently, adding those dependencies solves the problem.

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,64 @@
+        ('openwisp_users', '__first__'),
+        ('openwisp_users', '0004_default_groups'),
+        ('pki', '__first__'),
+        ('geo', '__first__'),
```
this shouldn't be necessary, try to remove it, if tests pass after you remove it don't add it back please

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,64 @@
+        try:
+            permission = Permission.objects.get(
+                codename="view_{}".format(modelClass)
+                )
```
same indentation style issue here

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,64 @@
+        for operation in manage_operations:
+            admin.permissions.add(
+                Permission.objects.get(codename="{}_{}".format(operation, modelClass)).pk
+                )
```
same indentation style issue here

nemesisdesign commented on openwisp_controller/config/migrations/0015_add_relevant_permissions_to_admin_and_operators.py:
```diff
@@ -0,0 +1,64 @@
+    operators_read_only_admins_manage = ['vpn']
+    manage_operations = ['add', 'change', 'delete']
+
+    for modelClass in operators_and_admins_can_change:
```
rename `modelClass` to `model_name`, don't use camel case notation for variables, camle case is reserved for classes.

Please apply this change to all other instances of `modelClass` in your code (or any other inappropiate use of camel case if present).

nemesisdesign commented on setup.py:
```diff
@@ -38,6 +38,7 @@
         "openwisp-utils[users]<0.3",
         "django-loci>=0.1.1,<0.3.0",
         "djangorestframework-gis>=0.12.0,<0.14.0"
+
```
please remove this blank line

nemesisdesign commented on tests/settings.py:
```diff
@@ -115,6 +115,8 @@
 EMAIL_PORT = '1025'  # for testing purposes
 LOGIN_REDIRECT_URL = 'admin:index'
 ACCOUNT_LOGOUT_REDIRECT_URL = LOGIN_REDIRECT_URL
+OPENWISP_ORGANIZATON_USER_ADMIN = True #tests will fail without this setting
```
good catch, this is something new in openwisp-users, but please respect PEP8, there should be 2 spaces before `#` and one after:
```python
OPENWISP_ORGANIZATON_USER_ADMIN = True  # tests will fail without this setting
```

shashwat1002 replied:
I didn't really catch it, @atb00ker  suggested :33
I'll take care of the spaces.

atb00ker replied:
@shashwat1002 this is still not solved! :(

nemesisdesign commented on tests/settings.py:
```diff
@@ -115,6 +115,8 @@
 LOGIN_REDIRECT_URL = 'admin:index'
 ACCOUNT_LOGOUT_REDIRECT_URL = LOGIN_REDIRECT_URL
+OPENWISP_ORGANIZATON_USER_ADMIN = True #tests will fail without this setting
+OPENWISP_ORGANIZATON_OWNER_ADMIN = True #tests will fail without this setting
```
same here
